integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanDatasetBuilder.java

```diff
@@ -0,0 +1,96 @@
+package io.openlineage.spark.api;
+
+import io.openlineage.client.OpenLineage;
+import io.openlineage.client.OpenLineage.Dataset;
+import io.openlineage.spark.agent.lifecycle.UnknownEntryFacetListener;
+import io.openlineage.spark.agent.util.ScalaConversionUtils;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.spark.scheduler.SparkListenerJobStart;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+
+/**
+ * {@link AbstractQueryPlanDatasetBuilder} serves as a bridge between the Abstract*DatasetBuilders
+ * and {@link QueryPlanVisitor}s. This base class provides the capability to trigger on a specific
+ * {@link org.apache.spark.scheduler.SparkListenerEvent} rather than visiting the query plan for
+ * every spark event (such as {@link org.apache.spark.scheduler.SparkListenerStageSubmitted}).
+ * Additionally, the triggering {@link org.apache.spark.scheduler.SparkListenerEvent} can be
+ * utilized during processing of the {@link LogicalPlan} nodes (e.g., reading Spark properties set
+ * in {@link SparkListenerJobStart#properties()}).
+ *
+ * @param <T>
+ * @param <P>
+ * @param <D>
+ */
+public abstract class AbstractQueryPlanDatasetBuilder<T, P extends LogicalPlan, D extends Dataset>
+    extends AbstractGenericArgPartialFunction<T, D> {
+  protected final OpenLineageContext context;
+  private final UnknownEntryFacetListener unknownEntryFacetListener =
+      UnknownEntryFacetListener.getInstance();
+
+  public AbstractQueryPlanDatasetBuilder(OpenLineageContext context) {
+    this.context = context;
+  }
+
+  protected DatasetFactory<OpenLineage.OutputDataset> outputDataset() {
+    return DatasetFactory.output(context.getOpenLineage());
+  }
+
+  protected DatasetFactory<OpenLineage.InputDataset> inputDataset() {
+    return DatasetFactory.input(context.getOpenLineage());
+  }
+
+  public abstract List<D> apply(P p);
+
+  public List<D> apply(T event) {
```

Comment:
When invoking partial functions, we invoke `io.openlineage.spark.agent.util.PlanUtils#merge` method which worked before as visitors where defined on disjunctive types. Now the types are common and mostly are just a couple of `SparkListenerEvent's.

I am having some issues when running integration tests for my branch: https://github.com/OpenLineage/OpenLineage/pull/573/files
and I think they may relate to this.

Reply:
Yeah, I need to finish up this change - it's incomplete and was already failing the tests. I took a look at your branch and I think I can finish up the changes in my branch and hopefully that addresses your failures. However, there are some compilation and formatting fixes that need to be made to your branch anyway. If you can get those addressed, I can re-run the tests with your branch rebased on mine after my fixes are done.